spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan resolveOperators {
+      case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
+          if matchedActions.isEmpty =>
+
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+        val targetTableScan = buildSimpleScanPlan(target.table, target.output, mergeBuilder, cond)
+
+        // when there are no matched actions, use a left anti join to remove any matching rows and rewrite to use
+        // append instead of replace. only unmatched source rows are passed to the merge and actions are all inserts.
+        val joinPlan = Join(source, targetTableScan, LeftAnti, Some(cond), JoinHint.NONE)
```

Comment:
Somewhat unrelated inquiry for my own understanding: If we pass `JoinHint.NONE`, are we allowing for any optimizations to take place? Join hints are what they seem to be, just the hints that we pass in either via the comments like `/* SKEW('field') */` or via the functions like `broadcast`, right?

In this case it likely doesn't matter much since this is being rewritten to an append, but I'm thinking if one of the tables is much smaller, a broadcast join might happen for example.

My two questions would be:
(1) In general, does passing in `JoinHint.NONE` stop spark from making its own calls on join types where it can using CBO etc?
(2) Given that Iceberg has all of its own stats handling, is there a situation where the join hints are going to be very useful?

Reply:
`JoinHint.NONE` is no hint, so Spark will use logic (e.g., CBO) other than the hint to decide whether to use a broadcast or sort-merge join.

In general, we don't want to do too much in Iceberg. Iceberg should provide stats for CBO and join logic in Spark, and the optimizations should be in Spark. There may be cases in the future where Spark is lagging, like bucketed joins, but ideally we will keep the logic in Spark.

Reply:
+1

Reply:
+1 on letting Spark do the estimation whether the table can be broadcasted. I am going to resolve this thread to simply the review. Please, reopen if needed.

Reply:
+1 thanks for the clarification and for the explanation on the thought process